embulk-core/src/main/java/org/embulk/spi/util/LineReader.java

```diff
@@ -0,0 +1,105 @@
+package org.embulk.spi.util;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.Reader;
+
+/**
+ * A {@link BufferedReader} that can specify line delimiter character from any one of CR, LF and CRLF.
+ * If not specified, follow the original {@link BufferedReader}'s behavior.
+ *
+ * This class is not thread-safe.
+ */
+class LineReader extends BufferedReader {
+    private static final int UNREAD = -1;
+    private final LineDelimiter lineDelimiter;
+    private final char[] buffer;
+    private int offset;
+    private int charsRead;
+
+    LineReader(Reader reader, LineDelimiter lineDelimiter, int bufferSize) {
```

Comment:
How about making this constructor `private`, and having another `static` initializer method (`.of` or `.create`, or else) to:

1. Return `BufferedReader` created as-is if `lineDelimiter` is `null`.
2. Return `LineReader` if `lineDelimiter` is not `null`.

So that we have less impacts by `LineReader` for existing configurations?

Reply:
Hmm, since current LineReader without "lineDelimiter" behaves completely same as BufferedReader, I don't think it makes less impact to current config.

Reply:
Sorry, I meant making it sure and clearer for those who read this code (readers). A simple `BufferedReader` instance returned is clearer for readers, rather than `LineReader` instance with `if (lineDelimiter == null)` check in `readLine`.

Reply:
Thanks, it's clear now and it makes sense. Fixed at 8317922